projection/core/src/main/scala/com/lightbend/lagom/internal/projection/ProjectionRegistryActor.scala

```diff
@@ -97,35 +107,51 @@ class ProjectionRegistryActor extends Actor with ActorLogging {
       log.debug(s"Registering worker $coordinates to [${sender().path.toString}]")
       // keep track
-      val workerKey: WorkerKey = coordinates.asKey
-      updateNameIndex(workerKey, coordinates)
-      actorIndex = actorIndex.updated(workerKey, sender())
-      reversedActorIndex = reversedActorIndex.updated(sender, workerKey)
+      actorIndex = actorIndex.updated(coordinates, sender())
+      reversedActorIndex = reversedActorIndex.updated(sender, coordinates)
       // when worker registers, we must reply with the requested status (if it's been set already, or DefaultInitialStatus if not).
-      val initialStatus = requestedStatusLocalCopy.getOrElse(workerKey, DefaultRequestedStatus)
-      log.debug(s"Setting initial status [$initialStatus] on worker $workerKey [${sender().path.toString}]")
+      val initialStatus = requestedStatusLocalCopy.getOrElse(coordinates, DefaultRequestedStatus)
+      log.debug(s"Setting initial status [$initialStatus] on worker $coordinates [${sender().path.toString}]")
       sender ! initialStatus
 
       // watch
       context.watch(sender)
 
     case GetState =>
       sender ! State.fromReplicatedData(
-        nameIndexLocalCopy,
+        nameIndex,
         requestedStatusLocalCopy,
         observedStatusLocalCopy,
         DefaultRequestedStatus,
         Stopped // unless observed somewhere (and replicated), we consider a worker stopped.
       )
 
-    // StateRequestCommand come from `ProjectionRegistry` and contain a requested Status
-    case command: StateRequestCommand =>
-      // locate the target actor and send the request
+    case RegisterProjection(projectionName, tagNames) =>
+      log.debug(s"Registering projection $projectionName for tags $tagNames.")
+      nameIndex += (projectionName -> tagNames.map {
+        WorkerCoordinates(projectionName, _)
+      })
+      // If we have stashed requestsfor this projection name, unstash them:
+      unknownProjections.get(projectionName).foreach { requestedStatus =>
+        self ! ProjectionRequestCommand(projectionName, requestedStatus)
+        unknownProjections -= projectionName
+      }
+
+    // XyzRequestCommand's come from `ProjectionRegistry` and contain a requested Status
+    case command: WorkerRequestCommand =>
       log.debug(s"Propagating request $command.")
-      updateStateChangeRequests(command.coordinates.asKey, command.requestedStatus)
+      updateStateChangeRequests(command.coordinates, command.requestedStatus)
+    case command: ProjectionRequestCommand =>
+      log.debug(s"Propagating request $command.")
+      val projectionWorkers: Option[Set[WorkerCoordinates]] = nameIndex.get(command.projectionName)
+      projectionWorkers match {
+        case Some(workerSet) =>
+          workerSet.foreach(coordinates => updateStateChangeRequests(coordinates, command.requestedStatus))
+        case None => unknownProjections += (command.projectionName -> command.requestedStatus)
```

Comment:
Do we really need this? We are not doing the same for the workers and the reason being this is already being done by the CRDT. We call `updateStateChangeRequests` for each worker so the request gets registered. 

When the projection gets registered in that node and the local workers start to bootstrap they will read from the CRDT map and get the right value. 

We are not really using the requested status for the projection. When we receive a `ProjectionRequestCommand` we use it to send `WorkerRequestCommand` to the workers.

Reply:
This is the bread and butter of this PR. This is the reason this whole change started in the first place: because we can only keep track of the worker-level requests in the CRDT, if we want to track projection-level requests that arrive _before_ a projection has registered then we need to stash those requests. I renamed `unknownProjections` to `unknownProjectionsStash` so the usage is clearer.

How `unknownProjections` is this: when a projection-level request comes in,

 - if the projection name is known, then we also know all the workers it has and we can expand the projection-level request to a collection of worker-level requests.
 - if the projection name is _not_ known, we stash the projection Name. Then, in `case RegisterProjection()` where details for new projections is handled, we check the stash for each incoming projection and expand the projection-level request to a collection of worker-level requests.

This way it doesn't matter what comes first: the projection registration of the projection-level request.

With this implementation the behavior for worker-level requests and projection-level request is identical:

1. registered or not, a request is tracked (unregarding the level)
2. it is no longer possible to get an unknown worker reporting for duty. 

Note that `2.` is a game changer that triggered a huge simplification. From `ProjectionRegistry`:


```diff 
 private[lagom] def registerProjection(...): Unit = {
+    projectionRegistryRef ! ProjectionRegistryActor.RegisterProjection(projectionName, shardNames)
    clusterShardingSettings.withRole(runInRole)
    clusterDistribution.start(...  )
  }
```

because all the details of a projection are now registered before any local or remote `EnsureActive` message for that projection can be handled it is now safe to assume a worker reporting for duty be known.

Having said all that:

> Do we really need this? We are not doing the same for the workers and the reason being this is already being done by the CRDT. We call updateStateChangeRequests for each worker so the request gets registered.

See above. We need it to support the projection-to-worker expansion of requests that arrived before projection registration.

> When the projection gets registered in that node and the local workers start to bootstrap they will read from the CRDT map and get the right value.

Only if stashing is used.

> We are not really using the requested status for the projection. When we receive a ProjectionRequestCommand we use it to send WorkerRequestCommand to the workers.

That is correct. projection-level and worker-level requests are now tracked in a single CRDT where the key is `WorkerCoordinates` and the value is `Status`. We no longer distinguish one or the other (see https://github.com/lagom/lagom/pull/2229/files#diff-0abf01cea9e7016c99ba6bb5a6ba179bR197) by type and only timing is considered, so `LastWriterWins` across all requests: sending a projection-level request invalidates any worker-level requests that we had on that projection.

Reply:
Correct, that's for the clarification. I have missed this.